perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -2896,120 +3104,883 @@ private void closeExtSourcesAfterSynchronization(ExtSource membersSource, ExtSou
+
+		String actualParentName;
+		try {
+			actualParentName = getGroupById(sess, groupToUpdate.getParentGroupId()).getShortName();
+		} catch (GroupNotExistsException e) {
+			throw new InternalErrorException(e);
+		}
+
+		Group newParentGroup = null;
+
+		if((candidateGroup.getParentGroupName() == null)){
+			if(!actualParentName.equals(baseGroup.getShortName())){
+				newParentGroup = baseGroup;
+			}
+		}
+
+		else if(!actualParentName.equals(candidateGroup.getParentGroupName())){
+			if(!actualParentName.equals(baseGroup.getShortName())) {
+				newParentGroup = baseGroup;
+			}
+			List<Group> subGroups = getAllSubGroups(sess, baseGroup);
+			for (Group subGroup : subGroups) {
+				if (subGroup.getShortName().equals(candidateGroup.getParentGroupName())) {
+					newParentGroup = subGroup;
+					break;
+				}
+			}
+		}
+		return newParentGroup;
+	}
+
+	/**
+	 * Update group description.
+	 *
+	 *  1. Compare descriptions.
+	 * 		1.1 If they differ, set new description into the old one, update whole group and return true.
+	 * 		    Otherwise return false.
+	 *
+	 * Method is used by group structure synchronization.
+	 *
+	 * @param sess
+	 * @param groupWithOldDescription group with description which will be chaged
+	 * @param groupWithNewDescription candidate group with description which will replace the old one
+	 * @return true if description changed, false otherwise
+	 * @throws InternalErrorException
+	 */
+	private boolean updateGroupDescription(PerunSession sess, Group groupWithOldDescription, CandidateGroup groupWithNewDescription) throws InternalErrorException {
+		if(groupWithOldDescription.getDescription() != null) {
+			if(!groupWithOldDescription.getDescription().equals(groupWithNewDescription.getDescription())){
+				groupWithOldDescription.setDescription(groupWithNewDescription.getDescription());
+				groupsManagerImpl.updateGroup(sess, groupWithOldDescription);
+				return true;
+			}
+		} else if(groupWithNewDescription.getDescription() != null){
+			groupWithOldDescription.setDescription(groupWithNewDescription.getDescription());
+			groupsManagerImpl.updateGroup(sess, groupWithOldDescription);
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Add missing groups under base group in Perun
+	 *
+	 * for every candidateGroup:
+	 * 1. Specify where to add that group
+	 * 2. Create group.
+	 *
+	 * If some problem occurs, add candidateGroup to skippedGroups and skip it.
+	 *
+	 * Method is used by group structure synchronization.
+	 *
+	 * @param sess
+	 * @param baseGroup under which we will be synchronizing groups
+	 * @param candidateGroupsToAdd list of new groups (candidateGroups)
+	 *
+	 * @throws InternalErrorException if some internal error occurs
+	 */
+	private void addMissingGroupsWhileSynchronization(PerunSession sess, Group baseGroup, List<CandidateGroup> candidateGroupsToAdd, List<String> skippedGroups) throws InternalErrorException {
```

Comment:
This method probably doesn't work as expected when the order of received groups is bottom-top.

Reply:
This looks valid when update is used for every group after adding.